File: backend/schema/Stats.cs
namespace Schema;

public static class Stats
{
    /// <summary>
    /// Returns the WPM at each second, based on the percent of the phrase 
    /// that was completed within that second.
    /// </summary>
    public static List<float> GetRawWpmBySecond(List<KeyStroke> keystrokes)
    {
        if (keystrokes.Count == 0)
        {
            return new List<float>();
        }

        List<float> wpmBySecond = new();
        List<int> charCountBySecond = new();
        for (int i = 0; i < keystrokes.Count; i++)
        {
            int second = (int)Math.Floor(keystrokes[i].Time);

            if (second < 0)
            {
                throw new InvalidOperationException("Can't be sending neagtive keystrokes. Got: " + second);
            }

            while (charCountBySecond.Count <= second)
                charCountBySecond.Add(0);
            while (wpmBySecond.Count <= second)
                wpmBySecond.Add(0);

            charCountBySecond[second]++;
        }

        for (int i = 0; i < wpmBySecond.Count; i++)
        {
            if (charCountBySecond[i] == 0)
            {
                wpmBySecond[i] = 0;
                continue;
            }

            wpmBySecond[i] = GetWpm(charCountBySecond[i], 1);
        }

        for (int i = 2; i < wpmBySecond.Count; i++)
        {
            wpmBySecond[i] = (wpmBySecond[i - 2] + wpmBySecond[i - 1] + wpmBySecond[i]) / 3;
        }

        return wpmBySecond;
    }

    /// <summary>
    /// Returns the running WPM at each second, based on what percent of the
    /// phrase they had completed by that point.
    /// </summary>
    public static List<float> GetAggWpmBySecond(List<KeyStroke> keyStrokes)
    {
        if (keyStrokes.Count == 0)
        {
            return new List<float>();
        }

        List<float> progressionStack = new();
        for (int i = 0; i < keyStrokes.Count; i++)
        {
            if (keyStrokes[i].Character == "\b")
            {
                progressionStack.RemoveAt(progressionStack.Count - 1);
            }
            else
            {
                progressionStack.Add(keyStrokes[i].Time);
            }
        }

        List<float> aggWpmByCharacter = new();
        for (int i = 0; i < progressionStack.Count; i++)
        {
            aggWpmByCharacter.Add(GetWpm(i + 1, progressionStack[i]));
        }

        int target = 1;
        List<int> nearestIndexPriorWpmToSecondBounds = new();
        for (int i = 0; i < aggWpmByCharacter.Count; i++)
        {
            while (progressionStack[i] > target)
            {
                target += 1;
                nearestIndexPriorWpmToSecondBounds.Add(Math.Max(i - 1, 0));
            }
        }

        List<float> wpmBySecond = new();
        for (int i = 0; i < nearestIndexPriorWpmToSecondBounds.Count; i++)
        {
            int second = i + 1;
            int priorI = nearestIndexPriorWpmToSecondBounds[i];
            float prevVal = aggWpmByCharacter[priorI];
            float nextVal = aggWpmByCharacter[priorI + 1];
            float priorTime = progressionStack[priorI];
            float nextTime = progressionStack[priorI + 1];
            float timespan = nextTime - priorTime;
            float percentAlongTimespan = (second - priorTime) / timespan;
            float lerpedWpm = prevVal + (nextVal - prevVal) * percentAlongTimespan;
            lerpedWpm = Math.Max(lerpedWpm, 0);
            wpmBySecond.Add(lerpedWpm);
        }

        float finalWpm = aggWpmByCharacter[^1];
        wpmBySecond.Add(finalWpm);

        return wpmBySecond;
    }

    public static List<ErrorsAtTime> GetErrorCountByTime(List<KeyStroke> keyStrokes, string phrase)
    {
        if (keyStrokes.Count == 0)
        {
            return new List<ErrorsAtTime>();
        }

        List<ErrorsAtTime> errorCountByTime = new() { new ErrorsAtTime { Time = 0, ErrorCount = 0 } };
        Stack<char> typedStack = new();
        int errorCount = 0;

        foreach (KeyStroke stroke in keyStrokes)
        {
            if (stroke.Character == "\b")
            {
                if (typedStack.Count > 0)
                {
                    if (typedStack.Count > phrase.Length || typedStack.Peek() != phrase[typedStack.Count - 1])
                    {
                        errorCount--;
                        errorCountByTime.Add(new ErrorsAtTime
                        {
                            Time = stroke.Time,
                            ErrorCount = errorCount
                        });
                    }

                    typedStack.Pop();
                }
            }
            else
            {
                typedStack.Push(stroke.Character[0]);

                if (typedStack.Count > phrase.Length || typedStack.Peek() != phrase[typedStack.Count - 1])
                {
                    errorCount++;
                    errorCountByTime.Add(new ErrorsAtTime
                    {
                        Time = stroke.Time,
                        ErrorCount = errorCount
                    });
                }
            }
        }

        return errorCountByTime;
    }

    public static float CalculateAccuracy(IList<KeyStroke> strokes, string phrase)
    {
        int errors = GetErrorCount(strokes, phrase);
        return 1 - errors / (float)phrase.Length;
    }

    public static int GetErrorCount(IList<KeyStroke> keyStrokes, string phrase)
    {
        Stack<char> typedStack = new();
        int errorCount = 0;

        foreach (KeyStroke stroke in keyStrokes)
        {
            if (stroke.Character == "\b")
            {
                if (typedStack.Count > 0)
                {
                    typedStack.Pop();
                }
            }
            else
            {
                typedStack.Push(stroke.Character[0]);

                if (typedStack.Count > phrase.Length || typedStack.Peek() != phrase[typedStack.Count - 1])
                {
                    errorCount++;
                }
            }
        }

        return errorCount;
    }

    public static float GetWpm(IList<KeyStroke> keyStrokes)
    {
        if (keyStrokes.Count == 0)
        {
            return 0;
        }

        string parsed = ParseKeystrokes(keyStrokes);
        return GetWpm(parsed.Length, keyStrokes[^1].Time);
    }

    public static float GetWpm(int charCount, float time_s)
    {
        return charCount / time_s * 60 / 5;
    }

    public static float GetTime(IEnumerable<KeyStroke> keystrokes)
    {
        return keystrokes.Last().Time;
    }

    public static float WpmToTime(float wpm, int charCount)
    {
        return (charCount * 60) / (5 * wpm);
    }

    public static string ParseKeystrokes(IList<KeyStroke> keyStrokes)
    {
        Stack<char> wordStack = new();
        foreach (KeyStroke keyStroke in keyStrokes)
        {
            if (keyStroke.Character == "\b")
            {
                if (wordStack.Count > 0)
                {
                    wordStack.Pop();
                }
            }
            else
            {
                wordStack.Push(keyStroke.Character[0]);
            }
        }

        string typedWord = new(wordStack.Reverse().ToArray());
        return typedWord;
    }

}

File: typeracing.io/src/stats.ts
import { Timestamp } from "firebase/firestore";

export type KeyStroke = {
  character: string;
  time: Timestamp;
};

export type ErrorsAtTime = {
  time: number;
  errorCount: number;
};

export function getRawWpmBySecond(keystrokes: KeyStroke[]): number[] {
  if (keystrokes.length === 0) {
    return [];
  }

  const wpmBySecond: number[] = [];
  const charCountBySecond: number[] = [];

  for (const keystroke of keystrokes) {
    const second = keystroke.time.seconds;

    if (second < 0) {
      throw new Error(`Can't be sending negative keystrokes. Got: ${second}`);
    }

    while (charCountBySecond.length <= second) {
      charCountBySecond.push(0);
    }
    while (wpmBySecond.length <= second) {
      wpmBySecond.push(0);
    }

    charCountBySecond[second]++;
  }

  for (let i = 0; i < wpmBySecond.length; i++) {
    if (charCountBySecond[i] === 0) {
      wpmBySecond[i] = 0;
      continue;
    }

    wpmBySecond[i] = getWpm(charCountBySecond[i], 1);
  }

  for (let i = 2; i < wpmBySecond.length; i++) {
    wpmBySecond[i] =
      (wpmBySecond[i - 2] + wpmBySecond[i - 1] + wpmBySecond[i]) / 3;
  }

  return wpmBySecond;
}

export function getAggWpmBySecond(keyStrokes: KeyStroke[]): number[] {
  if (keyStrokes.length === 0) {
    return [];
  }

  const progressionStack: number[] = [];
  for (const stroke of keyStrokes) {
    if (stroke.character === "\b") {
      progressionStack.pop();
    } else {
      progressionStack.push(stroke.time.seconds);
    }
  }

  const aggWpmByCharacter: number[] = [];
  for (let i = 0; i < progressionStack.length; i++) {
    aggWpmByCharacter.push(getWpm(i + 1, progressionStack[i]));
  }

  let target = 1;
  const nearestIndexPriorWpmToSecondBounds: number[] = [];
  for (let i = 0; i < aggWpmByCharacter.length; i++) {
    while (progressionStack[i] > target) {
      target += 1;
      nearestIndexPriorWpmToSecondBounds.push(Math.max(i - 1, 0));
    }
  }

  const wpmBySecond: number[] = [];
  for (let i = 0; i < nearestIndexPriorWpmToSecondBounds.length; i++) {
    const second = i + 1;
    const priorI = nearestIndexPriorWpmToSecondBounds[i];
    const prevVal = aggWpmByCharacter[priorI];
    const nextVal = aggWpmByCharacter[priorI + 1];
    const priorTime = progressionStack[priorI];
    const nextTime = progressionStack[priorI + 1];
    const timespan = nextTime - priorTime;
    const percentAlongTimespan = (second - priorTime) / timespan;
    let lerpedWpm = prevVal + (nextVal - prevVal) * percentAlongTimespan;
    lerpedWpm = Math.max(lerpedWpm, 0);
    wpmBySecond.push(lerpedWpm);
  }

  const finalWpm = aggWpmByCharacter[aggWpmByCharacter.length - 1];
  wpmBySecond.push(finalWpm);

  return wpmBySecond;
}

export function getErrorCountByTime(
  keyStrokes: KeyStroke[],
  phrase: string
): ErrorsAtTime[] {
  if (keyStrokes.length === 0) {
    return [];
  }

  const errorCountByTime: ErrorsAtTime[] = [{ time: 0, errorCount: 0 }];
  const typedStack: string[] = [];
  let errorCount = 0;

  for (const stroke of keyStrokes) {
    if (stroke.character === "\b") {
      if (typedStack.length > 0) {
        if (
          typedStack.length > phrase.length ||
          typedStack[typedStack.length - 1] !== phrase[typedStack.length - 1]
        ) {
          errorCount--;
          errorCountByTime.push({
            time: stroke.time.seconds,
            errorCount: errorCount,
          });
        }
        typedStack.pop();
      }
    } else {
      typedStack.push(stroke.character);
      if (
        typedStack.length > phrase.length ||
        typedStack[typedStack.length - 1] !== phrase[typedStack.length - 1]
      ) {
        errorCount++;
        errorCountByTime.push({
          time: stroke.time.seconds,
          errorCount: errorCount,
        });
      }
    }
  }

  return errorCountByTime;
}

export function calculateAccuracy(
  strokes: KeyStroke[],
  phrase: string
): number {
  const errors = getErrorCount(strokes, phrase);
  return 1 - errors / phrase.length;
}

export function getErrorCount(keyStrokes: KeyStroke[], phrase: string): number {
  const typedStack: string[] = [];
  let errorCount = 0;

  for (const stroke of keyStrokes) {
    if (stroke.character === "\b") {
      if (typedStack.length > 0) {
        typedStack.pop();
      }
    } else {
      typedStack.push(stroke.character);
      if (
        typedStack.length > phrase.length ||
        typedStack[typedStack.length - 1] !== phrase[typedStack.length - 1]
      ) {
        errorCount++;
      }
    }
  }

  return errorCount;
}

export function getWpm(input: KeyStroke[] | number, time_s?: number): number {
  if (Array.isArray(input)) {
    if (input.length === 0) {
      return 0;
    }
    const parsed = parseKeystrokes(input);
    return getWpm(parsed.length, input[input.length - 1].time.seconds);
  } else {
    const charCount = input;
    return ((charCount / time_s!) * 60) / 5;
  }
}

export function getTime(keystrokes: KeyStroke[]): number {
  return keystrokes[keystrokes.length - 1].time.seconds;
}

export function wpmToTime(wpm: number, charCount: number): number {
  return (charCount * 60) / (5 * wpm);
}

export function parseKeystrokes(keyStrokes: KeyStroke[]): string {
  const wordStack: string[] = [];
  for (const keyStroke of keyStrokes) {
    if (keyStroke.character === "\b") {
      if (wordStack.length > 0) {
        wordStack.pop();
      }
    } else {
      wordStack.push(keyStroke.character);
    }
  }

  return wordStack.join("");
}
